File: packages/geosolve-collaboration/src/source.ts
// SPDX-License-Identifier: GPL-3.0-or-later
import type { TextEdit, TextLimits, TextRevision, TextSnapshot } from "./index.js";
import type { OperationId } from "./host.js";
import { decode, encode, unicode } from "./host-codec.js";

export interface TextHistoryHorizon {
  readonly generation:number;readonly discardedEvents:number;readonly oldestRevision:TextRevision|null;
}
export interface UserTextHistory {
  readonly undoCount:number;readonly redoCount:number;readonly canUndo:boolean;readonly canRedo:boolean;
  readonly undoUnavailable:string|null;readonly redoUnavailable:string|null;
  readonly horizon:TextHistoryHorizon;
}
export interface SourceHostConfiguration {
  readonly documentEpoch: string;
  readonly serverEpoch: string;
  readonly initialInput: string;
  readonly files: Readonly<Record<string,string>>;
  readonly limits?: TextLimits;
}
export interface AcceptedSource { readonly modelRevision:number; readonly acceptedInput:string; readonly files:Readonly<Record<string,string>> }
export interface SourceSnapshot {
  readonly sequence:number;
  readonly accepted:AcceptedSource;
  readonly working:TextSnapshot;
  readonly fileIds:Readonly<Record<string,string>>;
  readonly pendingNotices:readonly { readonly path:string; readonly acceptedRevision:number; readonly acceptedSourceDigest:string; readonly reason:string }[];
  readonly hasPendingStage:boolean;
  readonly needsRecovery:boolean;
}
export interface SourceEdit { readonly start:number; readonly end:number; readonly expected:string; readonly replacement:string }
export interface SourcePatch { readonly baseSourceDigest:string; readonly candidateSourceDigest:string; readonly edits:readonly SourceEdit[] }
export interface SourceCapture {
  readonly handle:string;
  /** Exact durable admission capture; restoration also needs the authenticated journal basis. */
  readonly captureJson:string;
  readonly acceptedBasis:AcceptedSource;
  readonly working:TextSnapshot;
  readonly fileIds:Readonly<Record<string,string>>;
}
export interface PreparedSource {
  readonly ticket:string;
  readonly acceptedBasis:AcceptedSource;
  readonly candidateFiles:Readonly<Record<string,string>>;
  readonly changedPaths:readonly string[];
}
export type SourceReconciliation =
  | { readonly kind:"captured"; readonly path:string }
  | { readonly kind:"reconciled"; readonly path:string; readonly patch:SourcePatch }
  | { readonly kind:"pending"; readonly path:string; readonly reason:string };
/** Candidate source envelope, not a text/model ACK. The host durably pairs model
 * publication with the authority terminal event before installing either stage.
 */
export interface SourceStage {
  readonly status:"staged";
  readonly stageId:string;
  readonly sequence:number;
  readonly checkpointJson:string;
  readonly accepted:AcceptedSource;
  readonly workingRevision:TextRevision;
}
export type PersistSourceStage = (stage:SourceStage)=>Promise<void>;
export interface SourceNativeHandle {
  snapshot():string;
  checkpoint():string;
  textCheckpoint():Uint8Array;
  textChangesSince(revisionJson:string):string;
  generateSyncMessage(peer:string):Uint8Array|undefined;
  forgetPeer(peer:string):void;
  stageTextSync(peer:string,message:Uint8Array,actor:Uint8Array):string;
  stageTextChanges(changesJson:string,actor:Uint8Array):string;
  stageHostEdits(expectedJson:string,editsJson:string):string;
  stageUserTextChanges(changesJson:string,actor:Uint8Array,operationJson:string):string;
  stageUserFileEdits(expectedJson:string,editsJson:string,operationJson:string):string;
  stageUserWorkingEdits(expectedJson:string,editsJson:string,operationJson:string):string;
  stageUserUndo(operationJson:string):string;
  stageUserRedo(operationJson:string):string;
  userHistory(userId:string):string;
  captureApply():string;
  restoreApplyCapture(captureJson:string,basisJson:string):string;
  applyNeedsRebase(capture:string):boolean;
  prepareCanvasUpdate(patchesJson:string):string;
  prepareApplyUpdate(capture:string,filesJson:string):string;
  releaseHandle(handle:string):boolean;
  stageValidatedPublication(ticket:string,acceptedInput:string,workingJson:string,reconciliationsJson:string):string;
  commitStage(id:string):string;
  failStage(id:string):void;
  discardUnpersistedStage(id:string):void;
  free():void;
}
export interface SourceWasmModule {
  default(options:{module_or_path:Uint8Array|URL|Response|WebAssembly.Module}):Promise<unknown>;
  TrustedSourceHost:{ new(configurationJson:string,actor:Uint8Array):SourceNativeHandle; restore(configurationJson:string,actor:Uint8Array,checkpointJson:string):SourceNativeHandle };
}
export interface SourceHostOptions {
  readonly configuration:SourceHostConfiguration;
  readonly actor:Uint8Array;
  readonly checkpointJson?:string;
  readonly wasmModule?:SourceWasmModule;
  readonly wasm?:Uint8Array|URL|Response|WebAssembly.Module;
}

/** Trusted server composition of accepted source and independent raw draft text.
 * Preparation/capture handles do not lock typing while external model jobs run.
 */
export class TrustedSourceHost {
  private disposed=false;
  private pending?:SourceStage;
  private readonly captures=new Set<SourceCapture>();
  private readonly prepared=new Set<PreparedSource>();
  constructor(private readonly native:SourceNativeHandle) {}
  snapshot():SourceSnapshot {this.live();return decode(this.native.snapshot());}
  checkpoint():string {this.live();return this.native.checkpoint();}
  textCheckpoint():Uint8Array {this.live();return this.native.textCheckpoint();}
  /** Read committed deltas without retaining per-client server handshakes. */
  textChangesSince(revision:TextRevision):{readonly sourceSequence:number;readonly workingRevision:TextRevision;readonly changes:readonly (readonly number[])[]} {
    this.live();return decode(this.native.textChangesSince(encode(revision)));
  }
  generateSyncMessage(peer:string):Uint8Array|undefined {this.live();unicode(peer);return this.native.generateSyncMessage(peer);}
  forgetPeer(peer:string):void {this.live();unicode(peer);this.native.forgetPeer(peer);}
  /** Host must bind actor to authenticated editor session before calling. */
  stageTextSync(peer:string,message:Uint8Array,actor:Uint8Array):SourceStage {
    this.live();unicode(peer);checkActor(actor);return this.retainStage(this.native.stageTextSync(peer,message,actor));
  }
  stageTextChanges(changes:readonly Uint8Array[],actor:Uint8Array):SourceStage {
    this.live();checkActor(actor);return this.retainStage(this.native.stageTextChanges(encode(changes.map(bytes=>Array.from(bytes))),actor));
  }
  /** Principal/operation comes from authenticated host admission. Native derives
   * character ownership from change bytes; no client inverse token is accepted. */
  stageUserTextChanges(changes:readonly Uint8Array[],actor:Uint8Array,operation:OperationId):SourceStage {
    this.live();checkActor(actor);return this.retainStage(this.native.stageUserTextChanges(encode(changes.map(bytes=>Array.from(bytes))),actor,encode(operation)));
  }
  stageUserFileEdits(edits:readonly TextEdit[],operation:OperationId,expected:TextRevision=this.snapshot().working.revision):SourceStage {
    this.live();return this.retainStage(this.native.stageUserFileEdits(encode(expected),encode(edits),encode(operation)));
  }
  /** Trusted host ordered mixed splices/file lifecycle as one durable contribution. */
  stageUserWorkingEdits(edits:readonly TextEdit[],operation:OperationId,expected:TextRevision=this.snapshot().working.revision):SourceStage {
    this.live();return this.retainStage(this.native.stageUserWorkingEdits(encode(expected),encode(edits),encode(operation)));
  }
  stageUserUndo(operation:OperationId):SourceStage {this.live();return this.retainStage(this.native.stageUserUndo(encode(operation)));}
  stageUserRedo(operation:OperationId):SourceStage {this.live();return this.retainStage(this.native.stageUserRedo(encode(operation)));}
  userHistory(userId:string):UserTextHistory {this.live();unicode(userId);return decode(this.native.userHistory(userId));}
  /** External mirror/lifecycle edits already ordered by the host gateway. */
  stageHostEdits(edits:readonly TextEdit[],expected:TextRevision=this.snapshot().working.revision):SourceStage {
    this.live();return this.retainStage(this.native.stageHostEdits(encode(expected),encode(edits)));
  }
  captureApply():SourceCapture {this.live();const capture=decode<SourceCapture>(this.native.captureApply());this.captures.add(capture);return capture;}
  /** basis must come from authenticated durable admission, never a client body. */
  restoreApplyCapture(captureJson:string,authenticatedBasis:AcceptedSource):SourceCapture {
    this.live();unicode(captureJson);const capture=decode<SourceCapture>(this.native.restoreApplyCapture(captureJson,encode(authenticatedBasis)));this.captures.add(capture);return capture;
  }
  applyNeedsRebase(capture:SourceCapture):boolean {this.captureOwned(capture);return this.native.applyNeedsRebase(capture.handle);}
  prepareCanvasUpdate(patches:readonly {readonly path:string;readonly patch:SourcePatch}[]):PreparedSource {
    this.live();return this.retainPrepared(this.native.prepareCanvasUpdate(encode(patches)));
  }
  prepareApplyUpdate(capture:SourceCapture,rebasedFiles:Readonly<Record<string,string>>):PreparedSource {
    this.captureOwned(capture);return this.retainPrepared(this.native.prepareApplyUpdate(capture.handle,encode(rebasedFiles)));
  }
  release(handle:SourceCapture|PreparedSource):void {
    this.live();
    if ("handle" in handle) {this.captureOwned(handle);this.native.releaseHandle(handle.handle);this.captures.delete(handle);}
    else {this.preparedOwned(handle);this.native.releaseHandle(handle.ticket);this.prepared.delete(handle);}
  }
  /** Only after independent engine validation of this exact ticket's candidate.
   * Reconciliation must refer to latest working heads after the model job finishes.
   */
  stageValidatedPublication(prepared:PreparedSource,acceptedInput:string,expectedWorking:TextRevision,reconciliations:readonly SourceReconciliation[]):SourceStage {
    this.preparedOwned(prepared);unicode(acceptedInput);
    return this.retainStage(this.native.stageValidatedPublication(prepared.ticket,acceptedInput,encode(expectedWorking),encode(reconciliations)));
  }
  /** Persist exact envelope and associated authority/model transaction first. */
  commitStage(stage:SourceStage):SourceSnapshot {
    this.stageOwned(stage);const snapshot=decode<SourceSnapshot>(this.native.commitStage(stage.stageId));this.pending=undefined;
    for (const prepared of this.prepared) if (prepared.acceptedBasis.modelRevision<snapshot.accepted.modelRevision) {
      this.native.releaseHandle(prepared.ticket);this.prepared.delete(prepared);
    }
    return snapshot;
  }
  failStage(stage:SourceStage):void {this.stageOwned(stage);this.native.failStage(stage.stageId);this.pending=undefined;this.captures.clear();this.prepared.clear();}
  /** Only before any filesystem append starts; uncertain writes require failStage. */
  discardUnpersistedStage(stage:SourceStage):void {this.stageOwned(stage);this.native.discardUnpersistedStage(stage.stageId);this.pending=undefined;}
  async receiveText(peer:string,message:Uint8Array,actor:Uint8Array,persist:PersistSourceStage):Promise<SourceSnapshot> {return this.persist(this.stageTextSync(peer,message,actor),persist);}
  async receiveTextChanges(changes:readonly Uint8Array[],actor:Uint8Array,persist:PersistSourceStage):Promise<SourceSnapshot> {return this.persist(this.stageTextChanges(changes,actor),persist);}
  async editWorking(edits:readonly TextEdit[],expected:TextRevision,persist:PersistSourceStage):Promise<SourceSnapshot> {return this.persist(this.stageHostEdits(edits,expected),persist);}
  dispose():void {
    if (!this.disposed) {if (this.pending) throw Error("Resolve pending source persistence before disposing");this.disposed=true;this.captures.clear();this.prepared.clear();this.native.free();}
  }
  private async persist(stage:SourceStage,persist:PersistSourceStage):Promise<SourceSnapshot> {
    try {await persist(stage);}catch(failure){this.failStage(stage);throw failure;}
    return this.commitStage(stage);
  }
  private retainStage(text:string):SourceStage {const stage=decode<SourceStage>(text);this.pending=stage;return stage;}
  private retainPrepared(text:string):PreparedSource {const prepared=decode<PreparedSource>(text);this.prepared.add(prepared);return prepared;}
  private live():void {if(this.disposed)throw Error("Source host has been disposed");}
  private stageOwned(stage:SourceStage):void {this.live();if(stage!==this.pending)throw Error("Unknown or stale source stage");}
  private captureOwned(capture:SourceCapture):void {this.live();if(!this.captures.has(capture))throw Error("Unknown or stale source capture");}
  private preparedOwned(prepared:PreparedSource):void {this.live();if(!this.prepared.has(prepared))throw Error("Unknown or stale source preparation");}
}
export async function createTrustedSourceHost(options:SourceHostOptions):Promise<TrustedSourceHost> {
  checkActor(options.actor);const configuration=encode(options.configuration);
  if(options.checkpointJson!==undefined)unicode(options.checkpointJson);
  const path="./wasm/geosolve_collaboration_wasm.js";
  const module:SourceWasmModule=options.wasmModule??await import(path);
  let bytes=options.wasm;
  if(!bytes){const url=new URL("./wasm/geosolve_collaboration_wasm_bg.wasm",import.meta.url);if(url.protocol==="file:"){const nodeFs="node:fs/promises";bytes=await(await import(nodeFs)).readFile(url);}else bytes=url;}
  await module.default({module_or_path:bytes!});
  return new TrustedSourceHost(options.checkpointJson===undefined?new module.TrustedSourceHost(configuration,options.actor):module.TrustedSourceHost.restore(configuration,options.actor,options.checkpointJson));
}
function checkActor(actor:Uint8Array):void {if(!(actor instanceof Uint8Array)||actor.length===0||actor.length>64)throw Error("Actor must contain 1..64 bytes");}

File: crates/geosolve-collaboration-wasm/tests/source.rs
// SPDX-License-Identifier: GPL-3.0-or-later
use geosolve_collaboration::{
    SharedTextDocument, SharedTextLimits,
    source_patch::{SourceEdit, SourcePatch, source_digest},
};
use geosolve_collaboration_wasm::TrustedSourceHost;
use serde_json::{Value, json};
fn config(process: &str) -> String {
    json!({"documentEpoch":"doc-1","serverEpoch":process,"initialInput":"initial-model","files":{"main.ts":"const width = 12;\n","helper.ts":"export const = (\n"}}).to_string()
}
fn host() -> TrustedSourceHost {
    TrustedSourceHost::new(&config("process-1"), b"server").unwrap()
}
fn decode(text: &str) -> Value {
    serde_json::from_str(text).unwrap()
}
fn snapshot(state: &TrustedSourceHost) -> Value {
    decode(&state.snapshot().unwrap())
}
fn width_patch() -> Value {
    let old = "const width = 12;\n";
    let new = "const width = 14;\n";
    serde_json::to_value(SourcePatch {
        base_source_digest: source_digest(old),
        candidate_source_digest: source_digest(new),
        edits: vec![SourceEdit {
            start: 14,
            end: 16,
            expected: "12".into(),
            replacement: "14".into(),
        }],
    })
    .unwrap()
}
fn edit(state: &mut TrustedSourceHost, edits: &Value) -> Value {
    let basis = snapshot(state)["working"]["revision"].to_string();
    let write = decode(&state.stage_host_edits(&basis, &edits.to_string()).unwrap());
    decode(
        &state
            .commit_stage(write["stageId"].as_str().unwrap())
            .unwrap(),
    )
}
#[test]
fn text_ack_staging_and_model_preparation_remain_independent() {
    let mut state = host();
    let prepared = decode(
        &state
            .prepare_canvas_update(&json!([{"path":"main.ts","patch":width_patch()}]).to_string())
            .unwrap(),
    );
    let original = snapshot(&state);
    let old_heads = original["working"]["revision"].to_string();
    let write=decode(&state.stage_host_edits(&old_heads,&json!([{"kind":"splice","path":"main.ts","start_utf16":0,"delete_utf16":17,"insert":"const width ="}]).to_string()).unwrap());
    assert_eq!(snapshot(&state)["working"], original["working"]);
    assert_eq!(snapshot(&state)["sequence"], 0);
    assert!(state.capture_apply().is_err());
    assert!(state.commit_stage("foreign").is_err());
    state
        .commit_stage(write["stageId"].as_str().unwrap())
        .unwrap();
    let invalid = snapshot(&state);
    let updates =
        json!([{"kind":"pending","path":"main.ts","reason":"incomplete target"}]).to_string();
    assert!(
        state
            .stage_validated_publication(
                prepared["ticket"].as_str().unwrap(),
                "model-14",
                &old_heads,
                &updates
            )
            .is_err()
    );
    let terminal = decode(
        &state
            .stage_validated_publication(
                prepared["ticket"].as_str().unwrap(),
                "model-14",
                &invalid["working"]["revision"].to_string(),
                &updates,
            )
            .unwrap(),
    );
    assert_eq!(snapshot(&state)["accepted"], original["accepted"]);
    let durable = terminal["checkpointJson"].as_str().unwrap().to_owned();
    state
        .commit_stage(terminal["stageId"].as_str().unwrap())
        .unwrap();
    assert_eq!(
        snapshot(&state)["accepted"]["files"]["main.ts"],
        "const width = 14;\n"
    );
    assert_eq!(snapshot(&state)["working"], invalid["working"]);
    assert_eq!(
        snapshot(&state)["pendingNotices"].as_array().unwrap().len(),
        1
    );
    let restored =
        TrustedSourceHost::restore(&config("process-2"), b"new-server", &durable).unwrap();
    assert_eq!(snapshot(&restored), snapshot(&state));
}
#[test]
fn durable_apply_capture_restores_old_heads_basis_and_ids_after_later_typing() {
    let mut state = host();
    edit(
        &mut state,
        &json!([{"kind":"splice","path":"main.ts","start_utf16":14,"delete_utf16":2,"insert":"14"}]),
    );
    let captured = decode(&state.capture_apply().unwrap());
    edit(
        &mut state,
        &json!([{"kind":"splice","path":"main.ts","start_utf16":14,"delete_utf16":2,"insert":"16"}]),
    );
    let mut restored = TrustedSourceHost::restore(
        &config("process-2"),
        b"new-server",
        &state.checkpoint().unwrap(),
    )
    .unwrap();
    assert!(
        restored
            .prepare_apply_update(
                captured["handle"].as_str().unwrap(),
                &captured["working"]["files"].to_string()
            )
            .is_err()
    );
    let restored_capture = decode(
        &restored
            .restore_apply_capture(
                captured["captureJson"].as_str().unwrap(),
                &captured["acceptedBasis"].to_string(),
            )
            .unwrap(),
    );
    assert_eq!(restored_capture["working"], captured["working"]);
    assert_eq!(restored_capture["fileIds"], captured["fileIds"]);
    let prepared = decode(
        &restored
            .prepare_apply_update(
                restored_capture["handle"].as_str().unwrap(),
                &restored_capture["working"]["files"].to_string(),
            )
            .unwrap(),
    );
    let terminal = decode(
        &restored
            .stage_validated_publication(
                prepared["ticket"].as_str().unwrap(),
                "model-14",
                &snapshot(&restored)["working"]["revision"].to_string(),
                &json!([{"kind":"captured","path":"main.ts"}]).to_string(),
            )
            .unwrap(),
    );
    restored
        .commit_stage(terminal["stageId"].as_str().unwrap())
        .unwrap();
    assert_eq!(
        snapshot(&restored)["accepted"]["files"]["main.ts"],
        "const width = 14;\n"
    );
    assert_eq!(
        snapshot(&restored)["working"]["files"]["main.ts"],
        "const width = 16;\n"
    );
    let mut forged = decode(captured["captureJson"].as_str().unwrap());
    forged["files"]["main.ts"] = json!("forged");
    assert!(
        restored
            .restore_apply_capture(&forged.to_string(), &captured["acceptedBasis"].to_string())
            .is_err()
    );
}
#[test]
fn authenticated_text_failure_and_uncertain_durability_retain_committed_state() {
    let mut state = host();
    let mut client = SharedTextDocument::load(
        &state.text_checkpoint(),
        b"client",
        SharedTextLimits::default(),
    )
    .unwrap();
    let basis = client.revision();
    client.splice("main.ts", 14, 2, "16").unwrap();
    let changes = serde_json::to_string(&client.changes_since(&basis).unwrap()).unwrap();
    let before = snapshot(&state);
    assert!(state.stage_text_changes(&changes, b"forged").is_err());
    assert_eq!(snapshot(&state), before);
    let write = decode(&state.stage_text_changes(&changes, b"client").unwrap());
    assert_eq!(snapshot(&state)["working"], before["working"]);
    let durable = write["checkpointJson"].as_str().unwrap().to_owned();
    state
        .fail_stage(write["stageId"].as_str().unwrap())
        .unwrap();
    assert_eq!(snapshot(&state)["needsRecovery"], true);
    assert!(state.capture_apply().is_err());
    assert!(state.stage_text_changes(&changes, b"client").is_err());
    let restored =
        TrustedSourceHost::restore(&config("new-process"), b"new-server", &durable).unwrap();
    assert_eq!(
        snapshot(&restored)["working"]["files"]["main.ts"],
        "const width = 16;\n"
    );
    assert_eq!(snapshot(&restored)["accepted"], before["accepted"]);
}
#[test]
fn bounded_capture_handles_release_and_foreign_envelopes_refuse() {
    let mut state = host();
    let mut handles = Vec::new();
    for _ in 0..32 {
        handles.push(
            decode(&state.capture_apply().unwrap())["handle"]
                .as_str()
                .unwrap()
                .to_owned(),
        );
    }
    assert!(state.capture_apply().is_err());
    state.release_handle(&handles[0]).unwrap();
    assert!(state.capture_apply().is_ok());
    assert!(state.apply_needs_rebase(&handles[0]).is_err());
    let mut foreign = decode(&state.checkpoint().unwrap());
    foreign["documentEpoch"] = json!("other");
    assert!(TrustedSourceHost::restore(&config("new"), b"new", &foreign.to_string()).is_err());
}

#[test]
fn restore_authenticates_inner_epoch_and_source_model_sequence_relationship() {
    let mut state = host();
    let prepared = decode(
        &state
            .prepare_canvas_update(&json!([{"path":"main.ts","patch":width_patch()}]).to_string())
            .unwrap(),
    );
    let terminal = decode(
        &state
            .stage_validated_publication(
                prepared["ticket"].as_str().unwrap(),
                "model-14",
                &snapshot(&state)["working"]["revision"].to_string(),
                &json!([{"kind":"reconciled","path":"main.ts","patch":width_patch()}]).to_string(),
            )
            .unwrap(),
    );
    state
        .commit_stage(terminal["stageId"].as_str().unwrap())
        .unwrap();
    let checkpoint = decode(&state.checkpoint().unwrap());
    let mut invalid_sequence = checkpoint.clone();
    invalid_sequence["sequence"] = json!(0);
    assert!(
        TrustedSourceHost::restore(&config("new"), b"new", &invalid_sequence.to_string()).is_err()
    );
    let mut invalid_epoch = checkpoint.clone();
    let mut source = decode(checkpoint["sourceJson"].as_str().unwrap());
    source["documentEpoch"] = json!("other");
    invalid_epoch["sourceJson"] = json!(source.to_string());
    assert!(
        TrustedSourceHost::restore(&config("new"), b"new", &invalid_epoch.to_string()).is_err()
    );
}

#[test]
fn known_unpersisted_source_stage_discards_without_poison_and_old_stage_cannot_reappear() {
    let mut state = host();
    let before = state.checkpoint().unwrap();
    let revision = snapshot(&state)["working"]["revision"].to_string();
    let edits =
        json!([{"kind":"splice","path":"main.ts","start_utf16":14,"delete_utf16":2,"insert":"14"}])
            .to_string();
    let first = decode(&state.stage_host_edits(&revision, &edits).unwrap());
    state
        .discard_unpersisted_stage(first["stageId"].as_str().unwrap())
        .unwrap();
    assert_eq!(state.checkpoint().unwrap(), before);
    assert_eq!(snapshot(&state)["needsRecovery"], false);
    let second = decode(&state.stage_host_edits(&revision, &edits).unwrap());
    assert_ne!(first["stageId"], second["stageId"]);
    assert!(
        state
            .commit_stage(first["stageId"].as_str().unwrap())
            .is_err()
    );
    state
        .fail_stage(second["stageId"].as_str().unwrap())
        .unwrap();
    assert!(
        state
            .discard_unpersisted_stage(second["stageId"].as_str().unwrap())
            .is_err()
    );
    assert_eq!(snapshot(&state)["needsRecovery"], true);
    assert_eq!(state.checkpoint().unwrap(), before);
}

fn user_operation(user: &str, id: &str) -> String {
    json!({"userId":user,"clientId":format!("tab-{user}"),"requestId":id}).to_string()
}
fn commit_user_stage(state: &mut TrustedSourceHost, write: &str) {
    let write = decode(write);
    state
        .commit_stage(write["stageId"].as_str().unwrap())
        .unwrap();
}
fn user_type(state: &mut TrustedSourceHost, user: &str, id: &str, insert: &str) {
    let mut client = SharedTextDocument::load(
        &state.text_checkpoint(),
        user.as_bytes(),
        SharedTextLimits::default(),
    )
    .unwrap();
    let basis = client.revision();
    client.splice("main.ts", 14, 2, insert).unwrap();
    let write = state
        .stage_user_text_changes(
            &json!(client.changes_since(&basis).unwrap()).to_string(),
            user.as_bytes(),
            &user_operation(user, id),
        )
        .unwrap();
    commit_user_stage(state, &write);
}
#[test]
fn personal_text_staging_is_durable_and_checked_after_restart_with_a_new_server_actor() {
    let mut state = host();
    user_type(&mut state, "alice", "first", "14");
    user_type(&mut state, "alice", "second", "16");
    let before = state.checkpoint().unwrap();
    let undo = state
        .stage_user_undo(&user_operation("alice", "undo-second"))
        .unwrap();
    assert_eq!(state.checkpoint().unwrap(), before);
    assert_eq!(
        decode(&state.user_history("alice").unwrap())["undoCount"],
        2
    );
    let first = decode(&undo);
    state
        .discard_unpersisted_stage(first["stageId"].as_str().unwrap())
        .unwrap();
    let retry = state
        .stage_user_undo(&user_operation("alice", "undo-second"))
        .unwrap();
    assert_ne!(decode(&retry)["stageId"], first["stageId"]);
    commit_user_stage(&mut state, &retry);
    let saved = state.checkpoint().unwrap();
    let mut state = TrustedSourceHost::restore(&config("new"), b"new-server", &saved).unwrap();
    assert_eq!(
        snapshot(&state)["working"]["files"]["main.ts"],
        "const width = 14;\n"
    );
    let undo = state
        .stage_user_undo(&user_operation("alice", "undo-first"))
        .unwrap();
    commit_user_stage(&mut state, &undo);
    assert_eq!(
        snapshot(&state)["working"]["files"]["main.ts"],
        "const width = 12;\n"
    );
    let saved = state.checkpoint().unwrap();
    let mut state = TrustedSourceHost::restore(&config("third"), b"third-server", &saved).unwrap();
    let redo = state
        .stage_user_redo(&user_operation("alice", "redo-first"))
        .unwrap();
    commit_user_stage(&mut state, &redo);
    assert_eq!(
        snapshot(&state)["working"]["files"]["main.ts"],
        "const width = 14;\n"
    );
    assert_eq!(snapshot(&state)["accepted"]["modelRevision"], 0);
}
#[test]
fn personal_file_lifecycle_and_foreign_same_value_ownership_share_one_source_timeline() {
    let mut state = host();
    user_type(&mut state, "alice", "type", "14");
    user_type(&mut state, "bob", "same", "14");
    assert!(
        state
            .stage_user_undo(&user_operation("alice", "blocked"))
            .is_err()
    );
    let undo = state
        .stage_user_undo(&user_operation("bob", "undo"))
        .unwrap();
    commit_user_stage(&mut state, &undo);
    let original = snapshot(&state)["fileIds"]["main.ts"].clone();
    let basis = snapshot(&state)["working"]["revision"].to_string();
    let write = state
        .stage_user_file_edits(
            &basis,
            &json!([{"kind":"remove_file","path":"main.ts"}]).to_string(),
            &user_operation("alice", "delete"),
        )
        .unwrap();
    commit_user_stage(&mut state, &write);
    let saved = state.checkpoint().unwrap();
    let mut state = TrustedSourceHost::restore(&config("new"), b"new-server", &saved).unwrap();
    let write = state
        .stage_user_undo(&user_operation("alice", "restore"))
        .unwrap();
    commit_user_stage(&mut state, &write);
    assert_ne!(snapshot(&state)["fileIds"]["main.ts"], original);
    let write = state
        .stage_user_undo(&user_operation("alice", "undo-type"))
        .unwrap();
    commit_user_stage(&mut state, &write);
    assert_eq!(
        snapshot(&state)["working"]["files"]["main.ts"],
        "const width = 12;\n"
    );
}
